memgpt/agent.py

```diff
@@ -765,7 +782,12 @@ def validate_json(user_message_text: str, raise_on_error: bool) -> str:
             # (if yes) Step 5: send the info on the function call and function response to LLM
             response_message = response.choices[0].message
             response_message.model_copy()  # TODO why are we copying here?
-            all_response_messages, heartbeat_request, function_failed = self._handle_ai_response(response_message)
+            all_response_messages, heartbeat_request, function_failed = self._handle_ai_response(
+                response_message,
+                # TODO this is kind of hacky, find a better way to handle this
+                # the only time we set up message creation ahead of time is when streaming is on
+                response_message_id=response.id if stream else None,
```

Comment:
@sarahwooders translation:

If we're streaming (tokens), then we want to create a `Message.id` ahead of time so that the chunks we return via the API (and via the client once we add support) has `id`s attached to them.

However, this all happens *before* the MemGPT agent logic loop that takes a `ChatCompletionResponse` as input (which is the final result of a stream, not the intermediate result).

So that means we need to modify `handle_ai_response` to (in the streaming case) accept a pre-generated `Message.id`, and use it when we create the `Message` objects inside of `handle_ai_response`.